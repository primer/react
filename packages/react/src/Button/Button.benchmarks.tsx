import React, {Profiler} from 'react'
import {createRoot, type Root} from 'react-dom/client'
import {Button} from '../Button'
import BaseStyles from '../BaseStyles'
import ThemeProvider from '../ThemeProvider'

declare global {
  interface Window {
    BenchmarkRunner: BenchmarkRunner
  }
}

interface BenchmarkRunner {
  addBenchmark(benchmark: Benchmark): void
  runBenchmarks(): Promise<
    Array<{
      benchmark: Benchmark
      result: BenchmarkResult
    }>
  >
}

class BenchmarkRunner implements BenchmarkRunner {
  benchmarks: Array<Benchmark> = []

  addBenchmark(benchmark: Benchmark): void {
    this.benchmarks.push(benchmark)
  }

  async runBenchmarks(): Promise<
    Array<{
      benchmark: Benchmark
      result: BenchmarkResult
    }>
  > {
    const results = []

    for (const benchmark of this.benchmarks) {
      const runs: Array<BenchmarkRunResult> = []

      for (let i = 0; i < 1000; i++) {
        await benchmark.setup()

        const run = await benchmark.run()
        runs.push(run)

        await benchmark.teardown()
      }

      const result: BenchmarkResult = {
        averageBaseDuration: average(runs, 'baseDuration'),
        averageActualDuration: average(runs, 'actualDuration'),
        medianBaseDuration: median(runs, 'baseDuration'),
        medianActualDuration: median(runs, 'actualDuration'),
        runs,
      }

      results.push({
        benchmark,
        result,
      })
    }

    return results
  }
}

function average<K extends string, T extends Record<K, number>>(series: Array<T>, field: K): number {
  let sum = 0

  for (const value of series) {
    sum += value[field]
  }

  return sum / series.length
}

function median<K extends string, T extends Record<K, number>>(series: Array<T>, field: K): number {
  const values: Array<number> = series.map(value => value[field]).sort()
  return values[Math.floor(values.length / 2)]
}

interface Benchmark {
  name: string
  setup(): Promise<void> | void
  run(): Promise<BenchmarkRunResult>
  teardown(): Promise<void> | void
}

interface BenchmarkResult {
  averageActualDuration: number
  averageBaseDuration: number
  medianActualDuration: number
  medianBaseDuration: number
  runs: Array<BenchmarkRunResult>
}

interface BenchmarkRunResult {
  phase: 'update' | 'mount' | 'nested-update'
  actualDuration: number
  baseDuration: number
}

window.BenchmarkRunner = new BenchmarkRunner()

benchmark('Button', ({render, onRender}) => {
  render(
    <ThemeProvider>
      <BaseStyles>
        <Profiler id="benchmark" onRender={onRender}>
          <Button>Test case</Button>
        </Profiler>
      </BaseStyles>
    </ThemeProvider>,
  )
})

function benchmark(
  name: string,
  fn: ({render, onRender}: {render: (element: JSX.Element) => void; onRender: React.ProfilerProps['onRender']}) => void,
) {
  let root: Root | null = null

  function setup() {
    const node = document.createElement('div')
    node.setAttribute('id', 'benchmark')
    document.body.appendChild(node)
    root = createRoot(node)
  }

  function teardown() {
    root?.unmount()
    const node = document.getElementById('benchmark')
    if (node) {
      node.remove()
    }
    root = null
  }

  function run(): Promise<BenchmarkRunResult> {
    return new Promise(resolve => {
      const onRender: React.ProfilerProps['onRender'] = (_id, phase, actualDuration, baseDuration) => {
        resolve({
          phase,
          actualDuration,
          baseDuration,
        })
      }
      const render = (element: JSX.Element) => {
        root?.render(element)
      }
      fn({
        render,
        onRender,
      })
    })
  }

  const benchmark: Benchmark = {
    name,
    setup,
    teardown,
    run,
  }

  window.BenchmarkRunner.addBenchmark(benchmark)
}
