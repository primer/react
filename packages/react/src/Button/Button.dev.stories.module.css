.MediavsColor {
  &:hover {
    color: var(--fgColor-accent, var(--color-accent-fg));
  }

  @media screen and (max-width: 768px) {
    color: var(--fgColor-attention, var(--color-attention-fg));
    background-color: var(--bgColor-muted, var(--color-canvas-subtle));
  }

  @media (min-width: 1440px) {
    color: var(--fgColor-danger, var(--color-danger-fg));
    background-color: var(--bgColor-muted, var(--color-canvas-subtle));
  }
}

.MediumRed {
  color: var(--fgColor-danger, var(--color-danger-fg));
  background-color: var(--bgColor-muted, var(--color-canvas-subtle));
}

.DoneFg {
  color: var(--fgColor-done, var(--color-done-fg));
}

.SmallInvisibleSquare {
  width: var(--base-size-32, 32px);
  height: var(--base-size-32, 32px);
}

.SmallInvisibleSquare:focus {
  outline: 0;
}

.SmallInvisibleSquare:focus > span {
  outline: 2px solid var(--focus-outlineColor, var(--borderColor-accent-emphasis));
  outline-offset: 2px;
}

.BlockWidthOverride {
  width: 320px;
}

.LargeFontWatch {
  font-size: var(--text-title-size-large, 32px);
}
