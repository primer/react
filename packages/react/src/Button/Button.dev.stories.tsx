import {SearchIcon, TriangleDownIcon, EyeIcon, IssueClosedIcon} from '@primer/octicons-react'
import React from 'react'
import {Button, IconButton} from '.'
import {default as Text} from '../Text'

export default {
  title: 'Components/Button/DevOnly',
}

export const InvisibleVariants = () => {
  const count = 4
  return (
    <div style={{display: 'flex', flexDirection: 'row', gap: '1rem'}}>
      <Button variant="invisible">Button</Button>
      <Button variant="invisible" leadingVisual={SearchIcon}>
        Button
      </Button>
      <Button variant="invisible" trailingAction={TriangleDownIcon}>
        Button
      </Button>
      <Button variant="primary" count={count}>
        Button
      </Button>
      <Button variant="invisible" leadingVisual={EyeIcon} count={count}>
        Button
      </Button>
      <Button variant="invisible" leadingVisual={EyeIcon} trailingAction={TriangleDownIcon} count={count}>
        Button
      </Button>
      <IconButton icon={TriangleDownIcon} variant="invisible" aria-label="Invisible" />
    </div>
  )
}

export const TestSxProp = () => {
  const count = 4
  return (
    <div style={{display: 'flex', flexDirection: 'row', gap: '1rem'}}>
      <Button
        size="medium"
        sx={{
          color: 'firebrick',
          backgroundColor: '#F6F8FA',
        }}
      >
        Medium Red
      </Button>
      <Button
        size="small"
        sx={{
          ':hover': {
            color: 'deepskyblue',
          },
          [`@media screen and (max-width: 768px)`]: {
            color: 'maroon',
            backgroundColor: '#F6F8FA',
          },
          '@media (min-width: 1440)': {
            color: 'firebrick',
            backgroundColor: '#F6F8FA',
          },
        }}
      >
        Red
      </Button>
      <Button variant="invisible" sx={{color: 'firebrick', backgroundColor: '#F6F8FA'}}>
        Invariant color overridden
      </Button>
      <Button leadingVisual={IssueClosedIcon} sx={{color: 'done.fg'}}>
        <Text sx={{color: 'fg.default'}}>Close issue</Text>
      </Button>
      <Button
        size="small"
        variant="invisible"
        sx={{
          width: 32,
          height: 32,
          '&:focus': {
            outline: 0,
            '& > span': {
              boxShadow: `inset 0 0 0 2px deeppink`,
            },
          },
        }}
      >
        Custom size
      </Button>
      <Button size="small" block variant="invisible" sx={{width: 320}}>
        Overriden Block
      </Button>
      <Button sx={{fontSize: 32}} count={count}>
        Watch
      </Button>
    </div>
  )
}
