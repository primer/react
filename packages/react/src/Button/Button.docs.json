{
  "id": "button",
  "name": "Button",
  "status": "alpha",
  "a11yReviewed": false,
  "stories": [],
  "importPath": "@primer/react",
  "props": [
    {
      "name": "alignContent",
      "type": "'start' | 'center'",
      "description": "Content alignment for when visuals are present."
    },
    {
      "name": "as",
      "type": "React.ElementType",
      "defaultValue": "'button'"
    },
    {
      "name": "block",
      "type": "boolean",
      "description": "Allow button width to fill its container."
    },
    {
      "name": "children",
      "required": true,
      "type": "React.ReactNode",
      "description": "The content of the button."
    },
    {
      "name": "count",
      "type": "number | string",
      "description": "For counter buttons, the number to display."
    },
    {
      "name": "disabled",
      "type": "boolean",
      "description": "Items that are disabled can not be clicked, selected, or navigated through."
    },
    {
      "name": "inactive",
      "type": "boolean",
      "description": "Whether the button looks visually disabled, but can still accept all the same interactions as an enabled button.\n This is intended to be used when a system error such as an outage prevents the button from performing its usual action.\n Inactive styles are slightly different from disabled styles because inactive buttons need to have an accessible color contrast ratio. This is because inactive buttons can have tooltips or perform an action such as opening a dialog explaining why it's inactive.\n If both `disabled` and `inactive` are true, `disabled` takes precedence."
    },
    {
      "name": "leadingIcon",
      "type": "React.ComponentType<OcticonProps>",
      "description": "An icon to display before the button text.",
      "deprecated": true
    },
    {
      "name": "leadingVisual",
      "type": "React.ElementType",
      "description": "A visual to display before the button text."
    },
    {
      "name": "ref",
      "type": "React.RefObject<HTMLButtonElement>"
    },
    {
      "name": "size",
      "type": "'small'\n| 'medium'\n| 'large'",
      "defaultValue": "'medium'"
    },
    {
      "name": "sx",
      "type": "SystemStyleObject"
    },
    {
      "name": "trailingIcon",
      "type": "React.ComponentType<OcticonProps>",
      "description": "An icon to display after the button text.",
      "deprecated": true
    },
    {
      "name": "trailingVisual",
      "type": "React.ElementType",
      "description": "A visual to display after the button text."
    },
    {
      "name": "variant",
      "type": "'default'\n| 'primary'\n| 'danger'\n| 'invisible'",
      "defaultValue": "'default'",
      "description": "Change the visual style of the button."
    },
    {
      "name": "labelWrap",
      "type": "boolean",
      "defaultValue": "false",
      "description": "Whether the button label should wrap to multiple lines of it is longer than the button width."
    }
  ],
  "passthrough": {
    "element": "button",
    "url": "https://developer.mozilla.org/en-US/docs/Web/HTML/Element/button#attributes"
  }
}
