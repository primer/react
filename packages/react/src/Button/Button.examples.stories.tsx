import React from 'react'
import type {Meta} from '@storybook/react'
import {Button} from '.'
import {DownloadIcon} from '@primer/octicons-react'
import {Banner} from '../drafts'
import {AriaStatus} from '../live-region'

const meta: Meta<typeof Button> = {
  title: 'Components/Button/Examples',
} as Meta<typeof Button>

export default meta

export const LoadingStatusAnnouncementSuccessful = () => {
  const [loading, setLoading] = React.useState(false)
  const [success, setSuccess] = React.useState(false)

  const resolveAction = async () => {
    setLoading(true)
    await new Promise(resolve => setTimeout(resolve, 1500))
    setLoading(false)

    return await true
  }

  const onClick = (resolveType: 'error' | 'success') => async () => {
    const actionResult = await resolveAction()

    if (resolveType === 'error') {
      setSuccess(!actionResult)
      return
    }

    setSuccess(actionResult)
  }

  return (
    <>
      <AriaStatus>{!loading && success ? 'Export completed' : null}</AriaStatus>
      <Button loading={loading} leadingVisual={DownloadIcon} onClick={onClick('success')}>
        Export (success)
      </Button>
    </>
  )
}

export const LoadingStatusAnnouncementError = () => {
  const [loading, setLoading] = React.useState(false)
  const [error, setError] = React.useState(false)

  const resolveAction = async () => {
    setLoading(true)
    await new Promise(resolve => setTimeout(resolve, 1500))
    setLoading(false)

    return await true
  }

  const onClick = (resolveType: 'error' | 'success') => async () => {
    const actionResult = await resolveAction()

    if (resolveType === 'error') {
      setError(actionResult)
      return
    }

    setError(!actionResult)
  }

  return (
    <>
      {!loading && error ? <Banner title="Export failed" variant="critical" /> : null}

      <Button loading={loading} leadingVisual={DownloadIcon} onClick={onClick('error')}>
        Export (error)
      </Button>
    </>
  )
}
