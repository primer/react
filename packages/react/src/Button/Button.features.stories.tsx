import {EyeIcon, TriangleDownIcon, HeartIcon} from '@primer/octicons-react'
import React, {useState} from 'react'
import {Button} from '.'
import {Stack} from '../Stack/Stack'
import Link from '../Link'
import {announce} from '@primer/live-region-element'

export default {
  title: 'Components/Button/Features',
}

export const Primary = () => <Button variant="primary">Primary</Button>

export const Danger = () => <Button variant="danger">Danger</Button>

export const Invisible = () => <Button variant="invisible">Invisible</Button>

export const LeadingVisual = () => <Button leadingVisual={HeartIcon}>Leading visual</Button>

export const TrailingVisual = () => <Button trailingVisual={EyeIcon}>Trailing visual</Button>

const AccessibilityNote = () => {
  {
    return (
      <>
        <p>
          <b>Accessibility note</b>: If a button is dynamically updated to communicate a change (e.g. an action was
          successful), please make sure that this is also properly communicated to screen reader users. This may not
          happen reliably without additional markup considerations. Make sure to choose an approach that is appropriate
          for your usecase.
        </p>
        <p>
          Learn more about at{' '}
          <Link href="https://github.com/github/accessibility/blob/8b300b36d8bca28fd5e3e70ffa077a6f8ee65c05/docs/wiki/screen-reader-testing/dynamically-updated-buttons-support-april-2024.md">
            Staff-only: Dynamically updated button labels
          </Link>
          .
        </p>
      </>
    )
  }
}
export const TrailingCounter = () => {
  const [count, setCount] = useState(0)
  const onClick = () => {
    setCount(count + 1)
    announce(`Watch ${count + 1}`)
  }
  return (
    <>
      <Button onClick={onClick} count={count}>
        Watch
      </Button>
      <AccessibilityNote />
      <p>In this example, a live region has been implemented to communicate the change.</p>
    </>
  )
}

export const TrailingCounterAllVariants = () => {
  const [count, setCount] = useState(0)
  const onClick = () => {
    setCount(count + 1)
    announce(`Watch ${count + 1}`)
  }
  return (
    <>
      <div style={{display: 'flex', flexDirection: 'row', gap: '1rem'}}>
        <Button onClick={onClick} count={count}>
          Watch
        </Button>
        <Button onClick={onClick} count={count}>
          Watch
        </Button>
        <Button onClick={onClick} count={count}>
          Watch
        </Button>
        <Button onClick={onClick} variant="primary" disabled count={count}>
          Watch
        </Button>
        <Button onClick={onClick} variant="danger" count={count}>
          Watch
        </Button>
        <Button onClick={onClick} variant="danger" disabled count={count}>
          Watch
        </Button>
        <Button onClick={onClick} variant="invisible" count={count}>
          Watch
        </Button>
        <Button onClick={onClick} variant="invisible" disabled count={count}>
          Watch
        </Button>
      </div>
      <AccessibilityNote />
      <p>In these examples, a live region has been implemented to communicate the change.</p>
    </>
  )
}

export const TrailingAction = () => <Button trailingAction={TriangleDownIcon}>Trailing action</Button>

export const Block = () => <Button block>Default</Button>

export const Disabled = () => (
  <div style={{display: 'flex', flexDirection: 'row', gap: '1rem'}}>
    <Button disabled>Default</Button>
    <Button variant="primary" disabled>
      Primary
    </Button>
    <Button variant="danger" disabled>
      Danger
    </Button>
    <Button variant="invisible" disabled>
      Invisible
    </Button>
  </div>
)

export const Inactive = () => (
  <div style={{display: 'flex', flexDirection: 'row', gap: '1rem'}}>
    <Button inactive>Default</Button>
    <Button variant="primary" inactive>
      Primary
    </Button>
    <Button variant="danger" inactive>
      Danger
    </Button>
    <Button variant="invisible" inactive>
      Invisible
    </Button>
  </div>
)

export const Small = () => <Button size="small">Default</Button>

export const Medium = () => <Button size="medium">Default</Button>

export const Large = () => <Button size="large">Default</Button>

export const LabelWrap = () => {
  return (
    <Stack style={{width: '200px'}}>
      <Button labelWrap>This button label will wrap if the label is too long</Button>
      <Button size="small" labelWrap>
        This small button label will wrap if the label is too long
      </Button>
      <Button size="large" labelWrap>
        This large button label will wrap if the label is too long
      </Button>
      <Button labelWrap leadingVisual={HeartIcon} trailingVisual={EyeIcon}>
        This button label will wrap if the label is too long
      </Button>
      <Button labelWrap trailingAction={TriangleDownIcon}>
        This button label will wrap if the label is too long
      </Button>
    </Stack>
  )
}
