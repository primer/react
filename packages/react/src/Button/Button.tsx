import React, {forwardRef} from 'react'
import type {ButtonProps} from './types'
import {ButtonBase} from './ButtonBase'
import type {ForwardRefComponent as PolymorphicForwardRefComponent} from '../utils/polymorphic'
import {defaultSxProp} from '../utils/defaultSxProp'
import type {BetterSystemStyleObject, CSSCustomProperties} from '../sx'

const ButtonComponent = forwardRef(({children, sx: sxProp = defaultSxProp, ...props}, forwardedRef): JSX.Element => {
  const {block, size, leadingVisual, trailingVisual, trailingAction} = props
  let sxStyles = sxProp
  const style: CSSCustomProperties = {}

  if (sxProp !== null && Object.keys(sxProp).length > 0) {
    sxStyles = generateCustomSxProp({block, size, leadingVisual, trailingVisual, trailingAction}, sxProp)

    // @ts-ignore sxProp can have color attribute
    const {color} = sxProp
    if (color) style['--button-color'] = color
  }

  return (
    <ButtonBase ref={forwardedRef} as="button" sx={sxStyles} style={style} type="button" {...props}>
      {children}
    </ButtonBase>
  )
}) as PolymorphicForwardRefComponent<'button', ButtonProps>

// This function is used to generate a custom cssSelector for the sxProp

// The usual sx prop can like this:
// sx={{
//   [`@media (max-width: 768px)`]: {
//     '& > ul': {
//       backgroundColor: 'deeppink',
//     },
//     '&:hover': {
//       backgroundColor: 'yellow',
//     },
//   },
//  '&:hover': {
//     backgroundColor: 'yellow',
//   },
//  '&': {
//  width : 320px
// }
// }}
//*
/* What we want for Button styles is this:
sx={{
//   [`@media (max-width: 768px)`]: {
//     '&[data-attribute="something"] > ul': {
//       backgroundColor: 'deeppink',
//     },
//     '&[data-attribute="something"]:hover': {
//       backgroundColor: 'yellow',
//     },
//   },
//  '&[data-attribute="something"]:hover': {
//     backgroundColor: 'yellow',
//   },
//  '&[data-attribute="something"]': {
//     width : 320px
//  }
// }}

// We need to make sure we append the customCSSSelector to the original class selector. i.e & - > &[data-attribute="Icon"][data-size="small"]
*/
export function generateCustomSxProp(
  props: Partial<Pick<ButtonProps, 'size' | 'block' | 'leadingVisual' | 'trailingVisual' | 'trailingAction'>>,
  providedSx: BetterSystemStyleObject,
) {
  // Possible data attributes: data-size, data-block, data-no-visuals
  const size = props.size && props.size !== 'medium' ? `[data-size="${props.size}"]` : '' // medium is a default size therefore it doesn't have a data attribute that used for styling
  const block = props.block ? `[data-block="block"]` : ''
  const noVisuals = props.leadingVisual || props.trailingVisual || props.trailingAction ? '' : '[data-no-visuals]'

  // this is a custom selector. We need to make sure we add the data attributes to the base css class (& -> &[data-attributename="value"]])
  const cssSelector = `&${size}${block}${noVisuals}` // &[data-size="small"][data-block="block"][data-no-visuals]

  const customSxProp: {
    [key: string]: BetterSystemStyleObject
  } = {}

  if (!providedSx) return customSxProp
  else {
    customSxProp[cssSelector] = providedSx
    return customSxProp
  }
}

ButtonComponent.displayName = 'Button'

export {ButtonComponent}
