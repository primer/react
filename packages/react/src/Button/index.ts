import {IconButton} from './IconButton'
import {LinkButton} from './LinkButton'
export type {ButtonProps, IconButtonProps} from './types'
export {IconButton, LinkButton}
export {ButtonComponent as Button} from './Button'
export {ButtonBase} from './ButtonBase'
export type {ButtonBaseProps} from './ButtonBase'
