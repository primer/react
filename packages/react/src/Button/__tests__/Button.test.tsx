import {SearchIcon, HeartIcon} from '@primer/octicons-react'
import {render, screen, fireEvent} from '@testing-library/react'
import axe from 'axe-core'
import React from 'react'
import {IconButton, Button} from '../../Button'
import {behavesAsComponent} from '../../utils/testing'

describe('Button', () => {
  behavesAsComponent({Component: Button, options: {skipSx: true}})

  it('renders a <button>', () => {
    const container = render(<Button>Default</Button>)
    const button = container.getByRole('button')
    expect(button.textContent).toEqual('Default')
  })

  it('should have no axe violations', async () => {
    const {container} = render(<Button>Click here</Button>)
    const results = await axe.run(container)
    expect(results).toHaveNoViolations()
  })

  it('preserves "onClick" prop', () => {
    const onClick = jest.fn()
    const container = render(<Button onClick={onClick}>Noop</Button>)
    const button = container.getByRole('button')
    fireEvent.click(button)
    expect(onClick).toHaveBeenCalledTimes(1)
  })

  it('respects block prop', () => {
    const container = render(<Button block>Block</Button>)
    const button = container.getByRole('button')
    expect(button).toMatchSnapshot()
  })

  it('respects the "disabled" prop', () => {
    const onClick = jest.fn()
    const container = render(
      <Button onClick={onClick} disabled>
        Disabled
      </Button>,
    )
    const button = container.getByRole('button')
    expect(button.hasAttribute('disabled')).toEqual(true)
    fireEvent.click(button)
    expect(onClick).toHaveBeenCalledTimes(0)
  })

  it('respects the small size prop', () => {
    const container = render(<Button size="small">Smol</Button>)
    const button = container.getByRole('button')
    expect(button).toMatchSnapshot()
  })

  it('respects the large size prop', () => {
    const container = render(<Button size="large">Smol</Button>)
    const button = container.getByRole('button')
    expect(button).toMatchSnapshot()
  })

  it('styles primary button appropriately', () => {
    const container = render(<Button variant="primary">Primary</Button>)
    const button = container.getByRole('button')
    expect(button).toMatchSnapshot()
  })

  it('styles invisible button appropriately', () => {
    const container = render(<Button variant="invisible">Invisible</Button>)
    const button = container.getByRole('button')
    expect(button).toMatchSnapshot()
  })

  it('styles danger button appropriately', () => {
    const container = render(<Button variant="danger">Danger</Button>)
    const button = container.getByRole('button')
    expect(button).toMatchSnapshot()
  })

  it('makes sure icon button has an aria-label', () => {
    const container = render(<IconButton icon={SearchIcon} aria-label="Search button" />)
    const IconOnlyButton = container.getByLabelText('Search button')
    expect(IconOnlyButton).toBeTruthy()
  })

  it('respects the alignContent prop', () => {
    const container = render(<Button alignContent="start">Align start</Button>)
    const button = container.getByRole('button')
    expect(button).toMatchSnapshot()
  })

  it('should render the leadingVisual prop before the button content', () => {
    render(
      <Button leadingVisual={() => <span data-testid="leadingVisual" />}>
        <span>content</span>
      </Button>,
    )

    expect(screen.getByTestId('leadingVisual')).toBeInTheDocument()

    const position = screen.getByText('content').compareDocumentPosition(screen.getByTestId('leadingVisual'))
    expect(position).toBe(Node.DOCUMENT_POSITION_PRECEDING)
  })

  it('should render the trailingVisual prop after the button content', () => {
    render(
      <Button trailingVisual={() => <span data-testid="trailingVisual" />}>
        <span>content</span>
      </Button>,
    )
    expect(screen.getByTestId('trailingVisual')).toBeInTheDocument()

    const position = screen.getByText('content').compareDocumentPosition(screen.getByTestId('trailingVisual'))
    expect(position).toBe(Node.DOCUMENT_POSITION_FOLLOWING)
  })

  it('should render tooltip on an icon button when unsafeDisableTooltip prop is passed as false', () => {
    const {getByRole, getByText} = render(
      <IconButton icon={HeartIcon} aria-label="Heart" unsafeDisableTooltip={false} />,
    )
    const triggerEL = getByRole('button')
    const tooltipEl = getByText('Heart')
    expect(triggerEL).toHaveAttribute('aria-labelledby', tooltipEl.id)
  })
  it('should render description type tooltip on an icon button when unsafeDisableTooltip prop is passed as false', () => {
    const {getByRole, getByText} = render(
      <IconButton icon={HeartIcon} aria-label="Heart" description="Love is all around" unsafeDisableTooltip={false} />,
    )
    const triggerEL = getByRole('button')
    expect(triggerEL).toHaveAttribute('aria-label', 'Heart')
    const tooltipEl = getByText('Love is all around')
    expect(triggerEL).toHaveAttribute('aria-describedby', tooltipEl.id)
  })
  it('should not render tooltip on an icon button by default', () => {
    const {getByRole} = render(<IconButton icon={HeartIcon} aria-label="Heart" />)
    const triggerEl = getByRole('button')
    expect(triggerEl).not.toHaveAttribute('aria-labelledby')
    expect(triggerEl).toHaveAttribute('aria-label', 'Heart')
  })
})
