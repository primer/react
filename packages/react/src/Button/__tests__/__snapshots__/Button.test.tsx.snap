// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button respects block prop 1`] = `
<button
  aria-describedby="test-button-loading-announcement"
  class="ButtonBase"
  data-block="block"
  data-loading="false"
  data-no-visuals="true"
  data-size="medium"
  data-variant="default"
  id="test-button"
  type="button"
>
  <span
    class="ButtonContent"
    data-align="center"
    data-component="buttonContent"
  >
    <span
      class="Label"
      data-component="text"
    >
      Block
    </span>
  </span>
</button>
`;

exports[`Button respects the alignContent prop 1`] = `
<button
  aria-describedby="test-button-loading-announcement"
  class="ButtonBase"
  data-loading="false"
  data-no-visuals="true"
  data-size="medium"
  data-variant="default"
  id="test-button"
  type="button"
>
  <span
    class="ButtonContent"
    data-align="start"
    data-component="buttonContent"
  >
    <span
      class="Label"
      data-component="text"
    >
      Align start
    </span>
  </span>
</button>
`;

exports[`Button respects the large size prop 1`] = `
<button
  aria-describedby="test-button-loading-announcement"
  class="ButtonBase"
  data-loading="false"
  data-no-visuals="true"
  data-size="large"
  data-variant="default"
  id="test-button"
  type="button"
>
  <span
    class="ButtonContent"
    data-align="center"
    data-component="buttonContent"
  >
    <span
      class="Label"
      data-component="text"
    >
      Smol
    </span>
  </span>
</button>
`;

exports[`Button respects the small size prop 1`] = `
<button
  aria-describedby="test-button-loading-announcement"
  class="ButtonBase"
  data-loading="false"
  data-no-visuals="true"
  data-size="small"
  data-variant="default"
  id="test-button"
  type="button"
>
  <span
    class="ButtonContent"
    data-align="center"
    data-component="buttonContent"
  >
    <span
      class="Label"
      data-component="text"
    >
      Smol
    </span>
  </span>
</button>
`;

exports[`Button styles danger button appropriately 1`] = `
<button
  aria-describedby="test-button-loading-announcement"
  class="ButtonBase"
  data-loading="false"
  data-no-visuals="true"
  data-size="medium"
  data-variant="danger"
  id="test-button"
  type="button"
>
  <span
    class="ButtonContent"
    data-align="center"
    data-component="buttonContent"
  >
    <span
      class="Label"
      data-component="text"
    >
      Danger
    </span>
  </span>
</button>
`;

exports[`Button styles invisible button appropriately 1`] = `
<button
  aria-describedby="test-button-loading-announcement"
  class="ButtonBase"
  data-loading="false"
  data-no-visuals="true"
  data-size="medium"
  data-variant="invisible"
  id="test-button"
  type="button"
>
  <span
    class="ButtonContent"
    data-align="center"
    data-component="buttonContent"
  >
    <span
      class="Label"
      data-component="text"
    >
      Invisible
    </span>
  </span>
</button>
`;

exports[`Button styles primary button appropriately 1`] = `
<button
  aria-describedby="test-button-loading-announcement"
  class="ButtonBase"
  data-loading="false"
  data-no-visuals="true"
  data-size="medium"
  data-variant="primary"
  id="test-button"
  type="button"
>
  <span
    class="ButtonContent"
    data-align="center"
    data-component="buttonContent"
  >
    <span
      class="Label"
      data-component="text"
    >
      Primary
    </span>
  </span>
</button>
`;
